### ISSUE TITLE
Matrix multiplication in NL objective functions

### ISSUE THREAD

**Issue by tyleransom** [JuliaOpt/JuMP.jl#477]
I'm trying to get JuMP to do nonlinear optimization via NLopt where I have a vector of variables that is multiplied by another vector in the objective function. However, it looks like matrix multiplication isn't allowed in NL optimization at the moment.

The example below illustrates what I am running into. I am solving maximum likelihood of one of the simplest statistical models around.

`myMLE = Model(solver=NLoptSolver(algorithm=:LN_NELDERMEAD))`
`@defVar(myMLE, b[1:6], start = 0.2)`
`@defVar(myMLE, s >=0.0, start = 1.0)`
`@setNLObjective(myMLE, Max, (n/2)*log(1/(2*pi*s^2))-sum{(Y[i]-X[i,:]*b[:])^2, i=1:n}/(2s^2))`
`solve(myMLE)`

The product `X[i,:]*b[:]` appears to be problematic because size is not allowed (see [ticket #403](https://github.com/JuliaOpt/JuMP.jl/issues/403)). Furthermore, if I break out the vector b into scalars b1,...,b6 and then substitute `X[i,1]*b1-...-X[i,6]*b6` for `X[i,:]*b[:]`, everything works fine.

**Comment by mlubin** [comment]
As you observed, there are no vector operations supported in ``@setNLObjective``. Your two options are to:
1. Write out the matrix-vector operation in scalar terms using ``sum{}``
2. Use the (unreleased) vector operation functionality as follows:
```
@defVar(myMLE, YminusXb[1:length(Y)])
@addConstraint(myMLE, YminusXb .== Y-X*b)
@setNLObjective(myMLE, Max, (n/2)*log(1/(2*pi*s^2))-sum{(YminusXb[i])^2, i=1:n}/(2s^2))
```
This formulation introduces equality constraints, which most NLopt solvers don't support. I'd recommend using Ipopt instead.

**Comment by tyleransom** [comment]
Thanks for the quick response! I like your first option, and I hadn't thought of using the `sum{}` formulation. Thanks!

**Comment by tyleransom** [comment]
A semi-related note: It might be good to add to the JuMP documentation that a vector of starting values needs to be assigned within an index (and not as an array) like so:
`@defVar(m, x[i=1:16], start = xstar[i])`
where `xstar` is the desired vector of values (e.g. if hot-start is desired).

**Comment by mlubin** [comment]
Actually now that we have a concept of a one-based array of variables, we could potentially remove this restriction.

**Comment by mlubin** [comment]
Closed in favor of #481
